drivers/timer/arcv2_timer0.c

```diff
@@ -205,10 +205,14 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 
 	/* Round delay up to next tick boundary */
 	delay = ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
-	last_load = delay;
 
-	timer0_limit_register_set(last_load - 1);
-	timer0_control_register_set(_ARC_V2_TMR_CTRL_NH | _ARC_V2_TMR_CTRL_IE);
+	if (last_load != delay) {
```

Comment:
I'm not sure I understand what's happening here.  The "delay" number here is a real time value indicating how long from the current time the next z_clock_announce() call is expected.  The last_load stores the delay from the last time this code path was taken.  Why are you comparing the two for equality?

I mean, imagine that someone sets a delay of 10k cycles (to expire at "T=10000").  Then 5k cycles later (at T=5000) the code decides to cancel that older timeout and set another 10k cycles in the future (i.e. it should expire at T=15000).  This will compare the two and... **NOT** change the hardware state.  So the next timeout will happen needlessly 5000 cycles too early.

That's not exactly wrong, but it's weird.  Either I'm misunderstanding something, or you just want to remove this if() block entirely.

Reply:
There is a case where z_clock_set_timeout can be called mulitple times in a tick.  Then last_load and delay can be the same, i.e, next tick. For example,   thread runs very fast,  the scheduler calls update_cahce(1) multiple times in a tick.

Reply:
for your case.

someone sets a delay of 10k cycles (to expire at "T=10000"). Then 5k cycles later (at T=5000) the code decides to cancel that older timeout and set another 10k cycles in the future (i.e. it should expire at T=15000). This will compare the two and... NOT change the hardware state. So the next timeout will happen needlessly 5000 cycles too early.
1. set 10K ,  last_load =  10k
2. At T=5000,  set another 10K,   delay = 10k + 5K = 15K,  last_load = l0k,   delay != last_load  -> reprogram the tmer.

Reply:
> The "delay" number here is a real time value indicating how long from the current time the next z_clock_announce() call is expected

Indeed, I don't think that is an issue: elapsed() is added to the delay argument so it becomes a 'different' delay.